### PR TITLE
[WIP] Fix Spack on Python3.8

### PR DESCRIPTION
**Pull request by DiegoMagdaleno**
Most of this code is by @scheibelp however, his patches no longer work because huge changes happened to spack, this changes update some of those patches, spack still doesn't work and I plan on looking more into it, if someone could help me a little bit or the community would be appreciated.

Credit: https://github.com/spack/spack/pull/14187/files

### CODE REVIEW

**Comment by DiegoMagdaleno** [comment]
As of now the current error is 

```==> Warning: clang@11.0.3-apple cannot build optimized binaries for "skylake". Using best target possible: "x86_64"
==> Error: object.__init__() takes exactly one argument (the instance to initialize)```

**Comment by adamjstewart** [comment]
Once we get everything working, we should port the GitHub build actions to use Python 3.8 instead of Python 3.7. @alalazo does GitHub actions support Python 3.8 yet?

File: bin/spack [hunk 1]
```diff
@@ -62,3 +62,7 @@ if 'ruamel' in sys.modules:
 # Once we've set up the system path, run the spack main method
 import spack.main  # noqa
 sys.exit(spack.main.main())
+
+# Once we've set up the system path, run the spack main method
+if __name__ == "__main__":
+    sys.exit(spack.main.main())
```
Comment (adamjstewart) [inline bin/spack]: This does not seem to be the same as the logic in @scheibelp's PR

Comment (DiegoMagdaleno) [reply bin/spack]: https://github.com/spack/spack/blob/897700465bff418a414a9dbd2c12381b0b1cabfa/bin/spack

Comment (adamjstewart) [reply bin/spack]: Yes, your PR has `sys.exit(spack.main.main())` twice, his only has it once.

Comment (adamjstewart) [reply bin/spack]: Now you've undone his change altogether...

Comment (DiegoMagdaleno) [reply bin/spack]: Fixed, the error I get now seems to be related to a wrong initialization 

` ==> Error: object.__init__() takes exactly one argument (the instance to initialize) `

File: lib/spack/spack/fetch_strategy.py [hunk 2]
```diff
@@ -214,14 +214,16 @@ def mirror_id(self):
...
-    """Composite for a FetchStrategy object.
-
-    Implements the GoF composite pattern.
-    """
+    """Composite for a FetchStrategy object."""
+    
     matches = FetchStrategy.matches
-
+    
+    def __init__(self):
```
Comment (DiegoMagdaleno) [inline lib/spack/spack/fetch_strategy.py:222]: This line is very very suspicious I am going to check it out.

**Comment by michaelkuhn** [comment]
I have upgraded to Fedora 32, which uses Python 3.8.2, over the weekend and expected Spack to fail after seeing this PR. So far, everything seems to work fine, though. Anything I should be on the lookout for?

**Comment by alalazo** [comment]
@adamjstewart 

>  @alalazo does GitHub actions support Python 3.8 yet?

It should, at least according to [this](https://help.github.com/en/actions/language-and-framework-guides/using-python-with-github-actions)

**Comment by scheibelp** [comment]
> I have upgraded to Fedora 32, which uses Python 3.8.2, over the weekend and expected Spack to fail after seeing this PR. So far, everything seems to work fine, though. Anything I should be on the lookout for?

@michaelkuhn 

I think this will only affect Mac OS users. Specifically from https://www.python.org/downloads/release/python-383rc1/

> on macOS, the spawn start method is now used by default in multiprocessing
> pickle now uses Protocol 4 by default, improving performance

These two changes cause problems for Spack (the second is possibly OK, but the first is definitely problematic).

**Comment by michaelkuhn** [comment]
> @michaelkuhn
> I think this will only affect Mac OS users. Specifically from https://www.python.org/downloads/release/python-383rc1/

Ah, okay, got it. Lucky me. :laughing:

**Comment by DiegoMagdaleno** [comment]
@scheibelp Sadly I am a macOS user and this totally breaks Spack, hope it can get fixed, since I use Spack as my main package manager

**Comment by scheibelp** [comment]
> I am a macOS user and this totally breaks Spack, hope it can get fixed, since I use Spack as my main package manager

Apologies: I didn't mean to imply that this isn't important, I was just clarifying the issue in https://github.com/spack/spack/pull/16314#issuecomment-623608981 (this is useful for example if people want to reproduce the problem to test out fixes).

**Comment by DiegoMagdaleno** [comment]
Oh no my bad didn't mean to express it like you said it wasn't important, I am just saying that it would be nice to see a fix, I want to contribute somethings like Ruby build system and also some ruby packages, sadly for some things I need Python3.8 and I cant really use Spack as of know, I tried looking into the issue, but I cant get spack to tell me in what class it is failing.

**Pull request closed without merging**